ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java

```diff
@@ -877,30 +877,45 @@ private static String visitWithinExpression(final WithinExpressionContext contex
         final SqlBaseParser.SingleJoinWindowContext singleWithin
             = (SqlBaseParser.SingleJoinWindowContext) context;
 
-        stringBuilder.append(getSizeAndUnitFromJoinWindowSize(singleWithin.joinWindowSize()));
+        stringBuilder.append(String.format("%s",
+            anonymizeJoinWindowSize(singleWithin.joinWindowSize())));
+
+        if (singleWithin.gracePeriodClause() != null) {
+          stringBuilder.append(String.format(" GRACE PERIOD %s",
+              anonymizeGracePeriod(singleWithin.gracePeriodClause())));
+        }
       } else if (context instanceof JoinWindowWithBeforeAndAfterContext) {
         final SqlBaseParser.JoinWindowWithBeforeAndAfterContext beforeAndAfterJoinWindow
             = (SqlBaseParser.JoinWindowWithBeforeAndAfterContext) context;
 
-        final String before =
-            getSizeAndUnitFromJoinWindowSize(beforeAndAfterJoinWindow.joinWindowSize(0));
-        final String after =
-            getSizeAndUnitFromJoinWindowSize(beforeAndAfterJoinWindow.joinWindowSize(1));
-        stringBuilder.append(String.format("(%s, %s)", before, after));
+        stringBuilder.append(String.format("(%s, %s)",
+            anonymizeJoinWindowSize(beforeAndAfterJoinWindow.joinWindowSize(0)),
+            anonymizeJoinWindowSize(beforeAndAfterJoinWindow.joinWindowSize(1))));
+
+        if (beforeAndAfterJoinWindow.gracePeriodClause() != null) {
+          stringBuilder.append(String.format(" GRACE PERIOD %s",
+              anonymizeGracePeriod(beforeAndAfterJoinWindow.gracePeriodClause())));
+        }
       } else {
         throw new RuntimeException("Expecting either a single join window, ie \"WITHIN 10 "
-            + "seconds\", or a join window with before and after specified, "
-            + "ie. \"WITHIN (10 seconds, 20 seconds)");
+            + "seconds\" or \"WITHIN 10 seconds GRACE PERIOD 2 seconds\", or a join window with "
+            + "before and after specified, ie. \"WITHIN (10 seconds, 20 seconds)\" or "
+            + "WITHIN (10 seconds, 20 seconds) GRACE PERIOD 5 seconds");
       }
 
-      return stringBuilder.toString();
+      return stringBuilder.append(' ').toString();
```

Comment:
Is this change intentional? I assume it's fine if the tests below are passing, was just surprised to see it here.

Reply:
Yes. The previous string builders used to append the empty character at the end. To avoid duplicating the effort, I just moved that empty char to the end of the method.